src/audio/mux/mux_generic.c

```diff
@@ -18,347 +18,503 @@
+	uint32_t size_to_end; /* size to buffer end */
+	uint32_t min_frames;
+	uint32_t tmp_frames;
+	uint8_t elem;
+
+	comp_dbg(dev, "calc_frames_without_wrap(): frames: %d", frames);
+
+	min_frames = frames;
+
+	for (elem = 0; elem < lookup[0].num_elems; elem++) {
+		source = sources[lookup[0].copy_elem[elem].stream_id];
+		if (!source)
+			continue;
+
+		/* calculate sources min frames */
+		frame_bytes = audio_stream_frame_bytes(source);
+		sample_bytes = audio_stream_sample_bytes(source);
+		size_to_end = (char *)source->end_addr -
+			(char *)lookup[0].copy_elem[elem].src;
+		tmp_frames = size_to_end / frame_bytes;
+		/* check whether there is enough memory for tmp_frames and
+		 * one more channel
+		 */
+		if (size_to_end - (tmp_frames * frame_bytes) >= sample_bytes)
+			tmp_frames++;
+		min_frames = (tmp_frames < min_frames) ? tmp_frames :
+			min_frames;
+
+		/* calculate sink min frames */
+		frame_bytes = audio_stream_frame_bytes(sink);
+		sample_bytes = audio_stream_sample_bytes(sink);
+		size_to_end = (char *)sink->end_addr -
+			(char *)lookup[0].copy_elem[elem].dest;
+		tmp_frames = size_to_end / frame_bytes;
+		/* check whether there is enough memory for tmp_frames and
+		 * one more channel
+		 */
+		if (size_to_end - (tmp_frames * frame_bytes) >= sample_bytes)
+			tmp_frames++;
+		min_frames = (tmp_frames < min_frames) ? tmp_frames :
+			min_frames;
 	}
 
-	return sample;
+	return min_frames;
 }
 
-/* \brief Demuxing 16 bit streams.
- *
- * Source stream is routed to sink with regard to routing bitmasks from
- * mux_stream_data structure. Each bitmask describes composition of single
- * output channel.
- *
- * \param[in,out] sink Destination buffer.
- * \param[in,out] source Source buffer.
- * \param[in] frames Number of frames to process.
- * \param[in] data Parameters describing channel count and routing.
- */
-static void demux_s16le(struct audio_stream *sink,
-			const struct audio_stream *source, uint32_t frames,
-			struct mux_stream_data *data)
+static uint32_t demux_calc_frames_without_wrap(struct comp_dev *dev,
+					       struct audio_stream *sink,
+					       const struct audio_stream
+						*source,
+					       struct mux_look_up *lookup,
+					       uint32_t frames)
 {
-	int32_t sample;
-	int16_t *dst;
-	uint32_t dst_idx;
-	uint8_t i;
-	uint8_t out_ch;
-
-	for (i = 0; i < frames; i++) {
-		for (out_ch = 0; out_ch < sink->channels; out_ch++) {
-			sample = calc_sample_s16le(source,
-						   i * source->channels,
-						   data->mask[out_ch]);
-
-			/* saturate to 16 bits */
-			dst_idx = i * sink->channels + out_ch;
-			dst = audio_stream_write_frag_s16(sink, dst_idx);
-			*dst = sat_int16(sample);
-		}
+	uint32_t frame_bytes;
+	uint32_t sample_bytes;
+	uint32_t size_to_end; /* size to buffer end */
+	uint32_t min_frames;
+	uint32_t tmp_frames;
+	uint8_t elem;
+
+	comp_dbg(dev, "calc_frames_without_wrap(): frames: %d", frames);
+
+	min_frames = frames;
+
+	for (elem = 0; elem < lookup[0].num_elems; elem++) {
```

Comment:
Can you explain what this does inline. My expectation for demux is that source bytes would always be the same for each elem ? sink bytes would always be the same for mux ?

Reply:
@lgirdwood 
Do you mean: ```min_frames = frames```?
Firstly we calculate available frames in e.g. demux_copy():
```
for (i = 0; i < MUX_MAX_STREAMS; i++) {
		if (!sinks[i])
			continue;
		buffer_lock(sinks[i], &flags);
		avail = audio_stream_avail_frames(&source->stream,
						  &sinks[i]->stream);
		frames = MIN(frames, avail);
		buffer_unlock(sinks[i], flags);
	}
```
And we do not want in ```calc_frames_without_wrap()``` function, that frames_without_wrap will exceed available frames (frames_without_wrap calculation is based on buffer end address).